Port Playing Cards

RadsammyT opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
Ported playing cards to GoobStation from Frontier, from EstacaoPirata. Gamba is real now.

credits to Daniela for making the cards at EP. 
Leander-0 for porting to Frontier and whatson3 for making it good

You may obtain them during gameplay at a Games Vendor (3 in stock) or at lobby as a trinket in your loadout.

## Why / Balance
because i like gamba. and thought it would be amazing to see poker ops.

## Technical details
<!-- Summary of code changes for easier review. -->

Due to the structure of Frontier's port of the playing cards, I found it real easy to literally just drag-n-drop any folders with `EstacaoPirata` in them (sans reverting some frontier changes to its prototypes) and put them in their respective folders with it just working.

so while it is a working port, its prolly gonna be a shit one for reasons unknown to me. that and I don't know if its required to put all of this in `_Goobstation` folders rather than what its in now, so ¯\\_(ツ)\_/¯

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

https://github.com/user-attachments/assets/feb09348-6301-4fa6-be8f-f81dad920c38


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- add: Added Playing Cards. You may get them from the Games Vendor or as a trinket in your loadout.

Day-OS commented:
that might be the most awesome way of presenting this code I've seen yet! S2

I might add that the code has support for multilayered cards, so it is actually possible to port a complex card game like TGC into it, I'd love to see someone doing that haha!

whateverusername0 commented:
add mahjong next and i'm giving you all my goobcoins

whateverusername0 commented on Content.Client/_EstacaoPirata/Cards/CardSpriteSystem.cs:
```diff
@@ -0,0 +1,85 @@
+
+        var layerCount = 0;
+        //Gets the quantity of layers
+        foreach (var card in stack.Cards.TakeLast(cardCount))
+        {
+            if (!TryComp(card, out SpriteComponent? cardSprite))
+                return false;
+
+            layerCount += cardSprite.AllLayers.Count();
+        }
```
do what the ai said, the same thing caused void heretic to crash the server and i don't want that shit to repeat

whateverusername0 commented on Content.Server/_EstacaoPirata/OpenTriggeredStorageFill/OpenTriggeredStorageFillSystem.cs:
```diff
@@ -0,0 +1,53 @@
+    //Yes, that's a copy of StorageSystem StorageFill method
+    private void OnOpenEvent(EntityUid uid, OpenTriggeredStorageFillComponent comp, ActivateInWorldEvent args)
+    {
+        Log.Debug("aaa");
```
this

whateverusername0 commented on Content.Shared/_EstacaoPirata/Cards/Hand/CardHandSystem.cs:
```diff
@@ -0,0 +1,204 @@
+/// </summary>
+public sealed class CardHandSystem : EntitySystem
+{
+    public readonly EntProtoId CardHandBaseName = "CardHandBase";
```
add `[ValidatePrototypeId<EntityPrototype>]` attribute here just in case

RadsammyT replied:
done

RadsammyT replied:
all of these should be done
yell at me if it isnt

whateverusername0 reviewed (changes requested):
do whatever the bot said or tell the bot why he is wrong
also apply whatever changes i pointed to

whateverusername0 commented:
also playtest it thoroughly

Aviu00 commented on Content.Client/_EstacaoPirata/Cards/CardSpriteSystem.cs:
```diff
@@ -0,0 +1,85 @@
+        List<(int, ISpriteLayer)> layers = [];
+
+        var i = 0;
+        foreach (var card in stack.Cards.TakeLast(cardCount))
```
Should do the same thing here as the rabbit said above in `TryAdjustLayerQuantity` method.

RadsammyT replied:
done

Aviu00 reviewed (changes requested):
Also flipped cards should probably be hidden if a person holds them in their hands to prevent other people from seeing them in strip menu.

RadsammyT commented on Content.Shared/_EstacaoPirata/Cards/Stack/CardStackSystem.cs:
```diff
@@ -0,0 +1,453 @@
+        SubscribeLocalEvent<InteractUsingEvent>(OnInteractUsing);
+    }
+
+    public bool TryRemoveCard(EntityUid uid, EntityUid card, CardStackComponent? comp = null)
```
crash resolved. Shared CardStackSystem.OnActivate didn't actually set the handled argument when it was actually done

RadsammyT replied:
also I did somewhat implement the other thing you mentioned (auto 1-card convert) but for now it only happens when it its from an empty hand.
still need to figure out how to do the same thing on active items with `CardStackComponent`

RadsammyT replied:
done

RadsammyT commented on Content.Client/_EstacaoPirata/Cards/CardSpriteSystem.cs:
```diff
@@ -0,0 +1,87 @@
+
+            layerCount += cardSprite.AllLayers.Count();
+        }
+        layerCount = int.Max(1, layerCount); // Frontier: you need one layer.
```
int.Max is an actual method within dotnet runtime
https://learn.microsoft.com/en-us/dotnet/api/system.int32.max?view=net-9.0

RadsammyT commented on Content.Shared/_EstacaoPirata/Cards/Stack/CardStackSystem.cs:
```diff
@@ -0,0 +1,453 @@
+        else
+        {
+            if (firstComp.Cards.Count < MaxCardsInStack)
+                _audio.PlayPredicted(firstComp.PlaceDownSound, Transform(first).Coordinates, user);
```
this is in `JoinStacks` where its literally just a wrapper to `TryJoinStacks` just so it can be used as a verb. and the try version also does the audio calls. If this was called via verb then the possibility exists that we'd dupe sounds.

should be changed so that it just calls `TryJoinStacks`

RadsammyT commented on Content.Shared/_EstacaoPirata/Cards/Stack/CardStackSystem.cs:
```diff
@@ -0,0 +1,453 @@
+
+            if (soundUser != null)
+            {
+                _audio.PlayPredicted(firstComp.PlaceDownSound, Transform(firstStack).Coordinates, soundUser.Value);
```
should be changed so that its actually called on both.
and because of this `PlayPickupAnimation` has to be called on server otherwise shenanigans occur (if ran on both, animation dupes on client, etc.)

RadsammyT commented:
okay, it is currently just midnight and I am done addressing most of the reviews.
what remains is @Aviu00's review suggesting using `AppearanceSystem` rather than the original creator's method of constantly calling `UpdateSprite`.
which will be difficult for me because this is essentially the first time i've heard of it and will prolly see me refactoring the whole thing.
if anyone could help me with this, I will be very thankful and will give u all my goobcoin (~400)

Aviu00 commented:
> okay, it is currently just midnight and I am done addressing most of the reviews. what remains is @Aviu00's review suggesting using `AppearanceSystem` rather than the original creator's method of constantly calling `UpdateSprite`. which will be difficult for me because this is essentially the first time i've heard of it and will prolly see me refactoring the whole thing. if anyone could help me with this, I will be very thankful and will give u all my goobcoin (~400)

Actually, don't bother, it works currently for the most part. You'll only need to fix hand of cards always appearing as unflipped single card when client connects to the server. It works for deck of cards so just copy the logic from `CardDeckSystem` `Update` method to `CardHandSystem`, this might fix it.

Aviu00 commented:
And also, for example: when you take the second card from hand of cards with 2 cards in it, first card falls on the floor, they should both be in hands.

RadsammyT commented on Content.Client/_EstacaoPirata/Cards/Hand/CardHandSystem.cs:
```diff
@@ -0,0 +1,168 @@
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+        foreach (var (ent, value) in _notInit)
+        {
+            if (value >= 5)
+            {
+                _notInit.Remove(ent);
+                continue;
+            }
+            _notInit[ent] = value + 1;
+            if(!TryComp(ent.Owner, out CardStackComponent? stack) || stack.Cards.Count <= 0)
+                continue;
+            // If the card was STILL not initialized, we skip it
+            if (!TryGetCardLayer(stack.Cards.Last(), out var _))
+                continue;
+
+            // If cards were correctly initialized, we update the sprite
+            UpdateSprite(ent.Owner, ent.Comp);
+            _notInit.Remove(ent);
+        }
```
does this also apply to adding to the dictionary as well?

RadsammyT commented:
should be about everything done except for the hidden hands part. need to figure out how to get the hand's wielder uid (if any) and the clients own uid of the mob it controls.

or keep it as it is because its basically the only way you could "peek behind their shoulder" if you wanted to cheat or anything :godmode:

Aviu00 commented:
> should be about everything done except for the hidden hands part. need to figure out how to get the hand's wielder uid (if any) and the clients own uid of the mob it controls.

Edit `StrippableBoundUserInterface` and add extra check for cards in `AddInventoryButton` method.

Aviu00 commented on Content.Client/_EstacaoPirata/Cards/Hand/CardHandSystem.cs:
```diff
@@ -0,0 +1,168 @@
+            if(!TryComp(ent.Owner, out CardStackComponent? stack) || stack.Cards.Count <= 0)
+                continue;
+            // If the card was STILL not initialized, we skip it
+            if (!TryGetCardLayer(stack.Cards.Last(), out var _))
```
This needs to be slightly different from `CardDeckSystem`. This if statement can be removed I think.

RadsammyT replied:
if youre talking about the `TryGetCardLayer` if then alright. doesnt break anything from what I tested

Aviu00 commented on Content.Client/_EstacaoPirata/Cards/Hand/CardHandSystem.cs:
```diff
@@ -0,0 +1,168 @@
+        }
+    }
+
+    private bool TryGetCardLayer(EntityUid card, out SpriteComponent.Layer? layer)
```
Can be removed.

RadsammyT replied:
can't remove. made it as I was unable to call the one at `CardDeckSystem`.
the usage at `Update`, if removed, doesnt fuck things up but I cannot say the same for `UpdateSprite`. see below

Aviu00 commented on Content.Client/_EstacaoPirata/Cards/Hand/CardHandSystem.cs:
```diff
@@ -0,0 +1,168 @@
+        if (!TryComp(uid, out CardStackComponent? cardStack))
+            return;
+
+        // Prevents error appearing at spawnMenu
+        if (cardStack.Cards.Count <= 0 || !TryGetCardLayer(cardStack.Cards.Last(), out var cardlayer) ||
+            cardlayer == null)
+        {
+            _notInit[(uid, comp)] = 0;
+            return;
+        }
+
+        _cardSpriteSystem.TryAdjustLayerQuantity((uid, sprite, cardStack), comp.CardLimit);
+
+        var cardCount = Math.Min(cardStack.Cards.Count, comp.CardLimit);
+
+        // Frontier: zero/one card case
+        if (cardCount <= 0)
+        {
+            // Placeholder - we need to have a valid sprite.
+            sprite.LayerSetVisible(0, true);
+            sprite.LayerSetState(0, "singlecard_down_black");
+            sprite.LayerSetOffset(0, new Vector2(0f, 0f));
+            sprite.LayerSetScale(0, new Vector2(1f, 1f));
```
```suggestion
        _cardSpriteSystem.TryAdjustLayerQuantity((uid, sprite, cardStack), comp.CardLimit);

        var cardCount = Math.Min(cardStack.Cards.Count, comp.CardLimit);

        // Frontier: zero/one card case
        if (cardCount <= 0)
        {
            _notInit[(uid, comp)] = 0;
            // Placeholder - we need to have a valid sprite.
            sprite.LayerSetVisible(0, true);
            sprite.LayerSetState(0, "singlecard_down_black");
            sprite.LayerSetOffset(0, new Vector2(0f, 0f));
            sprite.LayerSetScale(0, new Vector2(1f, 1f));
```

RadsammyT replied:
doing this reintroduces the card-desync-on-join bug I believe.

Aviu00 commented:
If it doesn't work, then don't remove anything @RadsammyT

Aviu00 commented on Content.Client/Inventory/StrippableBoundUserInterface.cs:
```diff
@@ -171,8 +173,10 @@ private void AddHandButton(Hand hand)
-            UpdateEntityIcon(button, hand.HeldEntity);
+            //Goobstation: Cards are always hidden. NO CHEATING FOR U.
+            var isCard = EntMan.TryGetComponent<CardComponent>(hand.HeldEntity, out var _) ||
+                         EntMan.TryGetComponent<CardHandComponent>(hand.HeldEntity, out var _);
```
Use `HasComponent`.

Aviu00 commented on Content.Client/Inventory/StrippableBoundUserInterface.cs:
```diff
@@ -214,6 +218,15 @@ private void AddInventoryButton(EntityUid invUid, string slotId, InventoryCompon
+            // Goobstation: Playing Cards are always obscured in strip menu.
+            // I wanted to make the cards themselves appear hidden but this is simpler
+            var isCard = EntMan.TryGetComponent<CardComponent>(entity, out var _) ||
+                         EntMan.TryGetComponent<CardHandComponent>(entity, out var _);
```
Same here.
